driver-core/src/main/java/com/datastax/driver/core/Frame.java

```diff
@@ -201,19 +201,32 @@ protected void decode(ChannelHandlerContext ctx, ByteBuf buffer, List<Object> ou
             // version first bit is the "direction" of the frame (request or response)
             version = version & 0x7F;
 
-            DecoderForStreamIdSize decoder = (version >= 3) ? decoderV3 : decoderV1;
+            DecoderForStreamIdSize decoder = decoderForStreamIdSize.get();
+            if (decoder == null) {
```

Comment:
I think the only change required here is to remove the `static` modifier so that the decoders are not shared among connections.

The following makes your test pass as well:

```java
    static final class Decoder extends ByteToMessageDecoder {
        private final DecoderForStreamIdSize decoderV1 = new DecoderForStreamIdSize(1);
        private final DecoderForStreamIdSize decoderV3 = new DecoderForStreamIdSize(2);

        @Override
        protected void decode(ChannelHandlerContext ctx, ByteBuf buffer, List<Object> out) throws Exception {
            if (buffer.readableBytes() < 1)
                return;

            int version = buffer.getByte(0);
            // version first bit is the "direction" of the frame (request or response)
            version = version & 0x7F;

            DecoderForStreamIdSize decoder = (version >= 3) ? decoderV3 : decoderV1;
            Object frame = decoder.decode(ctx, buffer);
            if (frame != null)
                out.add(frame);
        }
...
```